lib/jwe.js

```diff
@@ -0,0 +1,511 @@
+  
+  return plainText;
+};
+
+
+/******************************************************************************
+ * JWE decryption
+ *****************************************************************************/
+
+/**
+ * decrypt cek helper function
+ *
+ * @param{String}  alg: algorithm
+ * @param{Buffer}  encrypted_cek: encrypted cek
+ * @param{Buffer}  key: encryption key
+ * @return{Function}  log: logging function
+ */
+var decryptCEKHelper = (alg, encrypted_cek, key, log) => {
+  var error = null;
+  var cek = null;
+
+  try {
+    var key_to_use;
+
+    if (alg === 'RSA1_5' || alg === 'RSA-OAEP') {
+      if (key['privatePemKey']) {
+        log.info('using RSA privatePemKey to decrypt cek');
+        key_to_use = key['privatePemKey'];
+      } else {
+        log.info('converting jwk to RSA privatePemKey to decrypt cek');
+        key_to_use = jwkToPem(key, {private: true});
+      }
+    } else {
+      log.info('using symmetric key to decrypt cek');
+      key_to_use = base64url.toBuffer(key.k);
+    }
+
+    if (alg === 'RSA1_5')
+      cek = crypto.privateDecrypt({ key: key_to_use, padding: constants.RSA_PKCS1_PADDING }, encrypted_cek);
+    else if (alg === 'RSA-OAEP')
+      cek = crypto.privateDecrypt({ key: key_to_use, padding: constants.RSA_PKCS1_OAEP_PADDING }, encrypted_cek);
+    else if (alg === 'A128KW')
+      cek = AESKeyUnWrap('aes-128-ecb', encrypted_cek, key_to_use);
+    else if (alg === 'A256KW')
+      cek = AESKeyUnWrap('aes-256-ecb', encrypted_cek, key_to_use);
+    else
+      cek = key_to_use;  // dir 
+  } catch (ex) {
+    error = ex;
+  }
+
+  return {'error': error, 'cek': cek};
+};
+
+/**
+ * decrypt cek function
+ *
+ * @param{Object}  header: header
+ * @param{Buffer}  encrypted_cek: encrypted cek
+ * @param{Object}  jweKeyStore: list of keys
+ * @return{Function}  log: logging function
+ */
+var decryptCEK = (header, encrypted_cek, jweKeyStore, log) => {
+  var algKtyMapper = { 'RSA1_5': 'RSA', 'RSA-OAEP': 'RSA', 'dir': 'oct', 'A128KW': 'oct', 'A256KW': 'oct'};
+
+  if (!header.alg)
+    return { 'error': new Error('alg is missing in JWE header'), 'cek': null };
+  if(['RSA1_5', 'RSA-OAEP', 'dir', 'A128KW', 'A256KW'].indexOf(header.alg) === -1)
+    return { 'error' : new Error('Unsupported alg in JWE header: ' + header.alg), 'cek': null };
+
+  var key = null;
+
+  if (header.kid) {
+    for (var i = 0; i < jweKeyStore.length; i++) {
+      if (header.kid === jweKeyStore[i].kid && algKtyMapper[header.alg] === jweKeyStore[i].kty) {
+        key = jweKeyStore[i];
+        log.info('found a key matching kid: ' + header.kid);
+        break;
+      }
+    }
+
+    if (key)
+      return decryptCEKHelper(header.alg, encrypted_cek, key, log);
+  }
+
+  log.info('we cannout find a key using kid, now we try all the possible keys');
+
+  // kid matching failed, now we try every possible key
+  for(var i = 0; i < jweKeyStore.length; i++) {
+    if (jweKeyStore[i].kty === algKtyMapper[header.alg]) {
```

Comment:
how does this work with 'dir', I am confused as the function is named 'decryptCEK' but with 'dir' the kek is not encrypted.